pyomo/contrib/alternative_solutions/solnpool.py

```diff
@@ -9,119 +9,1004 @@
+                tmp = []
+                for item in self._heap:
+                    value = -item.value if self.sense_is_min else item.value
+                    diff = (
+                        value - self.best_value
+                        if self.sense_is_min
+                        else self.best_value - value
+                    )
+                    if (
+                        (self.abs_tolerance is None) or (diff <= self.abs_tolerance)
+                    ) and (
+                        (self.rel_tolerance is None)
+                        or (
+                            diff / min(math.fabs(value), math.fabs(self.best_value))
+                            <= self.rel_tolerance
+                        )
+                    ):
+                        tmp.append(item)
+                    else:
+                        del self._solutions[item.id]
+                heapq.heapify(tmp)
+                self._heap = tmp
+
+            assert len(self._solutions) == len(
+                self._heap
+            ), f"Num solutions is {len(self._solutions)} but the heap size is {len(self._heap)}"
+            return soln.id
+
+        return None
+
+    def to_dict(self):
+        """
+        Converts SolutionPool to a dictionary object.
+
+        Returns
+        ----------
+        dict
+            Dictionary with three keys: 'metadata', 'solutions', 'pool_config'
+            'metadata' contains a dictionary of information about SolutionPools that is always present.
+            'solutions' contains a dictionary of the pool's solutions.
+            'pool_config' contains a dictionary of details conditional to the SolutionPool type.
+        """
+        return dict(
+            metadata=_to_dict(self.metadata),
+            solutions=_to_dict(self._solutions),
+            pool_config=_to_dict(self.pool_config),
+        )
+
+
+class PoolManager:
+    """
+    Manages one or more solution pools.
+
+    The default solution pool has policy ``keep_best`` with name ``None``.
+    If a new Solution pool is added without a name, then the ``None``
+    pool is replaced.  Otherwise, if a solution pool is added with an
+    existing name an error occurs.
+
+    The pool manager always has an active pool.  The pool manager has the
+    same API as a solution pool, and the envelope design pattern is used
+    to expose the methods and data for the active pool.  The active pool
+    defaults to the pool that was most recently added to the pool manager.
+
+    Note that all pools share the same Counter object to enable overall
+    solution count tracking and unique solution id values.
+
+    """
+
+    def __init__(self):
+        self._name = None
+        self._pools = {}
+        self.add_pool(name=self._name)
+        self._solution_counter = 0
+
+    #
+    # The following methods give the PoolManager the same API as a pool.
+    # These methods pass-though and operate on the active pool.
+    #
```

Comment:
Oh, I see, this is what "envelope design" is, right?

Reply:
Yes